File: src/types/site.ts
import z from "zod";

const enabledSchema = z.object({
  enabled: z.boolean(),
});

const additionalDomainSchema = z.object({
  domain: z.string(),
  redirect: enabledSchema,
  created_at: z.string(),
});

const httpsSchema = z.object({
  enabled: z.boolean(),
  certificate_expires: z.string().default(""),
  certificate_renews: z.string().default(""),
});

const nginxSchema = z.object({
  uploads_directory_protected: z.boolean(),
  xmlrpc_protected: z.boolean(),
  subdirectory_rewrite_in_place: z.boolean(),
});

const databaseSchema = z.object({
  id: z.number().nullish(),
  user_id: z.number().nullish(),
  table_prefix: z.string().nullish(),
});

const storageProviderSchema = z.object({
  id: z.number().nullish(),
  region: z.string().nullish(),
  bucket: z.string().nullish(),
});

const backupsSchema = z.object({
  files: z.boolean(),
  database: z.boolean(),
  paths_to_exclude: z.string().nullish(),
  is_backups_retention_period_enabled: z.boolean().nullish(),
  retention_period: z.number().nullish(),
  next_run_time: z.string().nullish(),
  storage_provider: storageProviderSchema.nullish(),
});

const basicAuthSchema = z.object({
  enabled: z.boolean(),
  username: z.string().nullish(),
});

export const siteSchema = z.object({
  id: z.number(),
  server_id: z.number(),
  domain: z.string().default(""),
  additional_domains: z.array(additionalDomainSchema),
  site_user: z.string().default(""),
  user_auth: z.string().default(""),
  php_version: z.string().default(""),
  public_folder: z.string().default(""),
  is_wordpress: z.boolean(),
  page_cache: enabledSchema,
  https: httpsSchema,
  nginx: nginxSchema,
  database: databaseSchema,
  backups: backupsSchema,
  wp_core_update: z.boolean(),
  wp_theme_updates: z.number().default(0),
  wp_plugin_updates: z.number().default(0),
  basic_auth: basicAuthSchema,
  created_at: z.string(),
  status: z.string(),
});

export type Site = z.infer<typeof siteSchema>;

export const cliSiteSchema = z.object({
  id: z.number(),
  name: z.string(),
  serverId: z.number(),
  serverName: z.string(),
  ssh: z.string(),
  path: z.string(),
});

export type CliSite = z.infer<typeof cliSiteSchema>;

File: internal/models/site.go
package models

type Enabled struct {
	Enabled bool `json:"enabled"`
}

type AdditionalDomain struct {
	Domain    string  `json:"domain"`
	Redirect  Enabled `json:"redirect"`
	CreatedAt string  `json:"created_at"`
}

type HTTPS struct {
	Enabled            bool   `json:"enabled"`
	CertificateExpires string `json:"certificate_expires"`
	CertificateRenews  string `json:"certificate_renews"`
}

type Nginx struct {
	UploadsDirectoryProtected  bool `json:"uploads_directory_protected"`
	XMLRPCProtected            bool `json:"xmlrpc_protected"`
	SubdirectoryRewriteInPlace bool `json:"subdirectory_rewrite_in_place"`
}

type SiteDatabase struct {
	ID          *int    `json:"id"`
	UserID      *int    `json:"user_id"`
	TablePrefix *string `json:"table_prefix"`
}

type StorageProvider struct {
	ID     *int    `json:"id"`
	Region *string `json:"region"`
	Bucket *string `json:"bucket"`
}

type Backups struct {
	Files                           bool             `json:"files"`
	Database                        bool             `json:"database"`
	PathsToExclude                  *string          `json:"paths_to_exclude"`
	IsBackupsRetentionPeriodEnabled *bool            `json:"is_backups_retention_period_enabled"`
	RetentionPeriod                 *int             `json:"retention_period"`
	NextRunTime                     *string          `json:"next_run_time"`
	StorageProvider                 *StorageProvider `json:"storage_provider"`
}

type BasicAuth struct {
	Enabled  bool    `json:"enabled"`
	Username *string `json:"username"`
}

type Site struct {
	ID                int                `json:"id"`
	ServerID          int                `json:"server_id"`
	Domain            string             `json:"domain"`
	AdditionalDomains []AdditionalDomain `json:"additional_domains"`
	SiteUser          string             `json:"site_user"`
	UserAuth          string             `json:"user_auth"`
	PHPVersion        string             `json:"php_version"`
	PublicFolder      string             `json:"public_folder"`
	IsWordpress       bool               `json:"is_wordpress"`
	PageCache         Enabled            `json:"page_cache"`
	HTTPS             HTTPS              `json:"https"`
	Nginx             Nginx              `json:"nginx"`
	Database          SiteDatabase       `json:"database"`
	Backups           Backups            `json:"backups"`
	WPCoreUpdate      bool               `json:"wp_core_update"`
	WPThemeUpdates    int                `json:"wp_theme_updates"`
	WPPluginUpdates   int                `json:"wp_plugin_updates"`
	BasicAuth         BasicAuth          `json:"basic_auth"`
	CreatedAt         string             `json:"created_at"`
	Status            string             `json:"status"`
}

type IgnoredSite struct {
	Domain    string `json:"domain"`
	Reason    string `json:"reason"`
	CreatedAt string `json:"created_at"`
}

type CliSite struct {
	ID         int    `json:"id"`
	Name       string `json:"name"`
	ServerID   int    `json:"serverId"`
	ServerName string `json:"serverName"`
	SSH        string `json:"ssh"`
	Path       string `json:"path"`
}
